bundle/src/main/java/com/adobe/acs/commons/wcm/impl/RobotsServlet.java

```diff
@@ -0,0 +1,418 @@
+        "sling.servlet.extensions=txt",
+        "sling.servlet.methods=GET"
+}, configurationPolicy = ConfigurationPolicy.REQUIRE)
+@Designate(ocd = RobotsServlet.RobotsServletConfig.class, factory = true)
+public final class RobotsServlet extends SlingSafeMethodsServlet {
+
+    private static final Logger log = LoggerFactory.getLogger(RobotsServlet.class);
+    private static final String ALLOW = "Allow: ";
+    private static final String USER_AGENT = "User-agent: ";
+    private static final String SITEMAP = "Sitemap: ";
+    private static final String DISALLOW = "Disallow: ";
+
+    private String externalizerDomain;
+
+    private String robotsContentsPropertyPath;
+
+    private boolean printGroupingComments;
+
+    private RobotsRuleSet rules;
+
+    @Reference
+    private Externalizer externalizer;
+
+    @Activate
+    protected void activate(RobotsServletConfig config) {
+        externalizerDomain = config.externalizer_domain();
+        robotsContentsPropertyPath = config.robots_content_property_path();
+        printGroupingComments = config.print_grouping_comments();
+
+        rules = new RobotsRuleSet(config);
+    }
+
+    @Override
+    protected void doGet(SlingHttpServletRequest request, SlingHttpServletResponse response)
+            throws ServletException, IOException {
+        response.setContentType("text/plain");
+        response.setCharacterEncoding("UTF-8");
+
+        write(request, response);
+    }
+
+    private void write(SlingHttpServletRequest request, SlingHttpServletResponse response) throws IOException {
+        if (StringUtils.isNotBlank(robotsContentsPropertyPath)) {
+            writeFromJcrProperty(request, response);
+        } else {
+            writeFromOsgiConfig(request, response);
+        }
+    }
+
+    private void writeFromOsgiConfig(SlingHttpServletRequest request, SlingHttpServletResponse response) throws IOException {
+        PrintWriter writer = response.getWriter();
+
+        rules.getGroups().forEach(group -> writeGroup(group, request, writer));
+
+        rules.getSitemaps().stream().map(sitemap -> buildSitemapString(sitemap, request.getResourceResolver())).forEach(writer::println);
+        if(!rules.getSitemapProperties().isEmpty()) {
+            PageManager pageManager = request.getResourceResolver().adaptTo(PageManager.class);
+            Page page = pageManager.getContainingPage(request.getResource());
+            addRulesFromPages(page, request.getResourceResolver(), rules.getSitemapProperties(), writer, this::buildSitemapString);
+        }
+    }
+
+    private void writeGroup(RobotsRuleGroup group, SlingHttpServletRequest request, PrintWriter writer) {
+        if (printGroupingComments) {
+            writer.println("# Start Group: " + group.getGroupName());
+        }
+
+        PageManager pageManager = request.getResourceResolver().adaptTo(PageManager.class);
+        Page page = pageManager.getContainingPage(request.getResource());
```

Comment:
Are we sure this is not null?

Reply:
See above for expected usage pattern. I suppose I can add null checks and return a 400 bad request if that somehow happens